File: tests/offsetImage.py
#!/usr/bin/env python2
from __future__ import absolute_import, division
from __future__ import print_function
from builtins import range

#
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.
#

"""
Tests for offsetting images in (dx, dy)

Run with:
   python offsetImage.py
or
   python
   >>> import offsetImage; offsetImage.run()
"""
import math

import unittest
import numpy

import lsst.utils.tests as utilsTests
import lsst.afw.image as afwImage
import lsst.afw.math as afwMath
import lsst.afw.geom as afwGeom
import lsst.afw.display.ds9 as ds9

try:
    type(display)
except NameError:
    display = False

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class offsetImageTestCase(unittest.TestCase):
    """A test case for offsetImage"""

    def setUp(self):
        self.inImage = afwImage.ImageF(200, 100)
        self.background = 200
        self.inImage.set(self.background)

    def tearDown(self):
        del self.inImage

    def testSetFluxConvervation(self):
        """Test that flux is preserved"""

        for algorithm in ("lanczos5", "bilinear", "nearest"):
            outImage = afwMath.offsetImage(self.inImage, 0, 0, algorithm)
            self.assertEqual(outImage.get(50, 50), self.background)

            outImage = afwMath.offsetImage(self.inImage, 0.5, 0, algorithm)
            self.assertAlmostEqual(outImage.get(50, 50), self.background, 4)

            outImage = afwMath.offsetImage(self.inImage, 0.5, 0.5, algorithm)
            self.assertAlmostEqual(outImage.get(50, 50), self.background, 4)

    def testSetIntegerOffset(self):
        """Test that we can offset by positive and negative amounts"""

        self.inImage.set(50, 50, 400)

        if False and display:
            frame = 0
            ds9.mtv(self.inImage, frame=frame)
            ds9.pan(50, 50, frame=frame)
            ds9.dot("+", 50, 50, frame=frame)

        for algorithm in ("lanczos5", "bilinear", "nearest"):
            for delta in [-0.49, 0.51]:
                for dx, dy in [(2, 3), (-2, 3), (-2, -3), (2, -3)]:
                    outImage = afwMath.offsetImage(self.inImage, dx + delta, dy + delta, algorithm)

                    if False and display:
                        frame += 1
                        ds9.mtv(outImage, frame=frame)
                        ds9.pan(50, 50, frame=frame)
                        ds9.dot("+", 50 + dx + delta - outImage.getX0(), 50 + dy + delta - outImage.getY0(),
                                frame=frame)

    def calcGaussian(self, im, x, y, amp, sigma1):
        """Insert a Gaussian into the image centered at (x, y)"""

        x = x - im.getX0()
        y = y - im.getY0()

        for ix in range(im.getWidth()):
            for iy in range(im.getHeight()):
                r2 = math.pow(x - ix, 2) + math.pow(y - iy, 2)
                val = math.exp(-r2/(2.0*pow(sigma1, 2)))
                im.set(ix, iy, amp*val)

    def testOffsetGaussian(self):
        """Insert a Gaussian, offset, and check the residuals"""

        size = 50
        refIm = afwImage.ImageF(size, size)
        unshiftedIm = afwImage.ImageF(size, size)

        xc, yc = size/2.0, size/2.0

        amp, sigma1 = 1.0, 3

        #
        # Calculate Gaussian directly at (xc, yc)
        #
        self.calcGaussian(refIm, xc, yc, amp, sigma1)

        for dx in (-55.5, -1.500001, -1.5, -1.499999, -1.00001, -1.0, -0.99999, -0.5,
            0.0, 0.5, 0.99999, 1.0, 1.00001, 1.499999, 1.5, 1.500001, 99.3):
            for dy in (-3.7, -1.500001, -1.5, -1.499999, -1.00001, -1.0, -0.99999, -0.5,
                0.0, 0.5, 0.99999, 1.0, 1.00001, 1.499999, 1.5, 1.500001, 2.99999):
                dOrigX, dOrigY, dFracX, dFracY = getOrigFracShift(dx, dy)
                self.calcGaussian(unshiftedIm, xc - dFracX, yc - dFracY, amp, sigma1)

                for algorithm, maxMean, maxLim in (
                    ("lanczos5", 1e-8, 0.0015),
                    ("bilinear", 1e-8, 0.03),
                    ("nearest",  1e-8, 0.2),
                ):
                    im = afwImage.ImageF(size, size)
                    im = afwMath.offsetImage(unshiftedIm, dx, dy, algorithm)


                    if display:
                        ds9.mtv(im, frame=0)

                    im -= refIm

                    if display:
                        ds9.mtv(im, frame=1)

                    imArr = im.getArray()
                    imGoodVals = numpy.ma.array(imArr, copy=False, mask=numpy.isnan(imArr)).compressed()

                    try:
                        imXY0 = tuple(im.getXY0())
                        self.assertEqual(imXY0, (dOrigX, dOrigY))
                        self.assertLess(abs(imGoodVals.mean()), maxMean*amp)
                        self.assertLess(abs(imGoodVals.max()), maxLim*amp)
                        self.assertLess(abs(imGoodVals.min()), maxLim*amp)
                    except:
                        print("failed on algorithm=%s; dx = %s; dy = %s" % (algorithm, dx, dy))
                        raise

# the following would be preferable if there was an easy way to NaN pixels
#
#         stats = afwMath.makeStatistics(im, afwMath.MEAN | afwMath.MAX | afwMath.MIN)
#
#         if not False:
#             print "mean = %g, min = %g, max = %g" % (stats.getValue(afwMath.MEAN),
#                                                      stats.getValue(afwMath.MIN),
#                                                      stats.getValue(afwMath.MAX))
#
#         self.assertTrue(abs(stats.getValue(afwMath.MEAN)) < 1e-7)
#         self.assertTrue(abs(stats.getValue(afwMath.MIN)) < 1.2e-3*amp)
#         self.assertTrue(abs(stats.getValue(afwMath.MAX)) < 1.2e-3*amp)


def getOrigFracShift(dx, dy):
    """Return the predicted integer shift to XY0 and the fractional shift that offsetImage will use

    offsetImage preserves the origin if dx and dy both < 1 pixel; larger shifts are to the nearest pixel.
    """
    if (abs(dx) < 1) and (abs(dy) < 1):
        return (0, 0, dx, dy)

    dOrigX = math.floor(dx + 0.5)
    dOrigY = math.floor(dy + 0.5)
    dFracX = dx - dOrigX
    dFracY = dy - dOrigY
    return (int(dOrigX), int(dOrigY), dFracX, dFracY)


class transformImageTestCase(unittest.TestCase):
    """A test case for rotating images"""

    def setUp(self):
        self.inImage = afwImage.ImageF(20, 10)
        self.inImage.set(0, 0, 100)
        self.inImage.set(10, 0, 50)

    def tearDown(self):
        del self.inImage

    def testRotate(self):
        """Test that we end up with the correct image after rotating by 90 degrees"""

        for nQuarter, x, y in [(0, 0, 0),
                               (1, 9, 0),
                               (2, 19, 9),
                               (3, 0, 19)]:
            outImage = afwMath.rotateImageBy90(self.inImage, nQuarter)
            if display:
                ds9.mtv(outImage, frame=nQuarter, title="out %d" % nQuarter)
            self.assertEqual(self.inImage.get(0, 0), outImage.get(x, y))

    def testFlip(self):
        """Test that we end up with the correct image after flipping it"""

        frame = 2
        for flipLR, flipTB, x, y in [(True, False, 19, 0),
                                     (True, True,  19, 9),
                                     (False, True, 0,  9),
                                     (False, False, 0, 0)]:
            outImage = afwMath.flipImage(self.inImage, flipLR, flipTB)
            if display:
                ds9.mtv(outImage, frame=frame, title="%s %s" % (flipLR, flipTB))
                frame += 1
            self.assertEqual(self.inImage.get(0, 0), outImage.get(x, y))

    def testMask(self):
        """Test that we can flip a Mask"""
        mask = afwImage.MaskU(10, 20)
        afwMath.flipImage(mask, True, False) # for a while, swig couldn't handle the resulting Mask::Ptr

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class binImageTestCase(unittest.TestCase):
    """A test case for binning images"""

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def testBin(self):
        """Test that we can bin images"""

        inImage = afwImage.ImageF(203, 131)
        inImage.set(1)
        bin = 4

        outImage = afwMath.binImage(inImage, bin)

        self.assertEqual(outImage.getWidth(), inImage.getWidth()//bin)
        self.assertEqual(outImage.getHeight(), inImage.getHeight()//bin)

        stats = afwMath.makeStatistics(outImage, afwMath.MAX | afwMath.MIN)
        self.assertEqual(stats.getValue(afwMath.MIN), 1)
        self.assertEqual(stats.getValue(afwMath.MAX), 1)

    def testBin2(self):
        """Test that we can bin images anisotropically"""

        inImage = afwImage.ImageF(203, 131)
        val = 1
        inImage.set(val)
        binX, binY = 2, 4

        outImage = afwMath.binImage(inImage, binX, binY)

        self.assertEqual(outImage.getWidth(), inImage.getWidth()//binX)
        self.assertEqual(outImage.getHeight(), inImage.getHeight()//binY)

        stats = afwMath.makeStatistics(outImage, afwMath.MAX | afwMath.MIN)
        self.assertEqual(stats.getValue(afwMath.MIN), val)
        self.assertEqual(stats.getValue(afwMath.MAX), val)

        inImage.set(0)
        subImg = inImage.Factory(inImage, afwGeom.BoxI(afwGeom.PointI(4, 4), afwGeom.ExtentI(4, 8)),
                                 afwImage.LOCAL)
        subImg.set(100)
        del subImg
        outImage = afwMath.binImage(inImage, binX, binY)

        if display:
            ds9.mtv(inImage, frame=2, title="unbinned")
            ds9.mtv(outImage, frame=3, title="binned %dx%d" % (binX, binY))

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""

    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(offsetImageTestCase)
    suites += unittest.makeSuite(transformImageTestCase)
    suites += unittest.makeSuite(binImageTestCase)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)
    return unittest.TestSuite(suites)

def run(shouldExit=False):
    """Run the tests"""
    utilsTests.run(suite(), shouldExit)

if __name__ == "__main__":
    run(True)

File: tests/offsetImage.cc
// -*- LSST-C++ -*-
//
// A machine-precision test of offsetImage(), complementary to the tests in offsetImage.py
//
// TODO: only bilinear interpolation is fully tested for now; should add lanczos and nearest-neighbor

/* 
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 * 
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the LSST License Statement and 
 * the GNU General Public License along with this program.  If not, 
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
 
#include <iostream>
#include <boost/random.hpp>

#include "lsst/afw/geom.h"
#include "lsst/afw/math.h"
#include "lsst/afw/image.h"
#include "lsst/afw/detection.h"

using namespace std;
using namespace boost;
using namespace lsst::afw::geom;
using namespace lsst::afw::math;
using namespace lsst::afw::image;


static random::mt19937 rng(0);  // RNG deliberately initialized with same seed every time
static random::uniform_int_distribution<> uni_int(0,100);
static random::uniform_01<> uni_double;


static void fill_random(Image<double> &im)
{
    int nx = im.getWidth();
    int ny = im.getHeight();

    double sum = 0.0;
    for (int i = 0; i < nx; i++) {
	for (int j = 0; j < ny; j++) {
	    im(i,j) = uni_double(rng);
	    sum += im(i,j);
	}
    }

    // image must be normalized to sum 1, due to current confusion in Psf::computeImage()
    for (int i = 0; i < nx; i++)
	for (int j = 0; j < ny; j++)
	    im(i,j) /= sum;
}


static inline void show_image(const Image<double> &im)
{
    int nx = im.getWidth();
    int ny = im.getHeight();

    cerr << "(x0,y0)=(" << im.getX0() << "," << im.getY0() << ")\n";
    for (int i = 0; i < nx; i++) {
	for (int j = 0; j < ny; j++)
	    cerr << " " << im(i,j);
	cerr << endl;
    }
}


// -------------------------------------------------------------------------------------------------
//
// testFlipOffset(): offsetting an image by dx and (-dx) should be related by the appropriate flip symmetry


static PTR(Image<double>) xflip(const Image<double> &im)
{
    int nx = im.getWidth();
    int ny = im.getHeight();
   
    PTR(Image<double>) ret = make_shared<Image<double> >(nx,ny);
    ret->setXY0(-im.getX0()-nx+1, im.getY0());

    for (int i = 0; i < nx; i++)
	for (int j = 0; j < ny; j++)
	    (*ret)(i,j) = im(nx-1-i,j);

    return ret;
}


static PTR(Image<double>) yflip(const Image<double> &im)
{
    int nx = im.getWidth();
    int ny = im.getHeight();
   
    PTR(Image<double>) ret = make_shared<Image<double> >(nx,ny);
    ret->setXY0(im.getX0(), -im.getY0()-ny+1);

    for (int i = 0; i < nx; i++)
	for (int j = 0; j < ny; j++)
	    (*ret)(i,j) = im(i,ny-1-j);

    return ret;
}


// returns 0 on success, 1 on failure
static int testOneFlipOffset(int nx, int ny, int x0, int y0, double dx, double dy, const string &algorithmName, unsigned int buffer)
{
    // identifier string which will be displayed on failure
    ostringstream os;
    os << "testOneFlipOffset(" << nx << "," << ny << "," << x0 << "," << y0 << "," << dx << "," << dy << "," << algorithmName << "," << buffer << ")";
    string s = os.str();

    // initialize PSF to random image
    Image<double> src(nx, ny);
    src.setXY0(x0, y0);
    fill_random(src);

    Image<double>::Ptr dst = offsetImage(src, dx, dy, algorithmName, buffer);

    Image<double>::Ptr dst2 = xflip(src);
    dst2 = offsetImage(*dst2, -dx, dy, algorithmName, buffer);
    dst2 = xflip(*dst2);

    Image<double>::Ptr dst3 = yflip(src);
    dst3 = offsetImage(*dst3, dx, -dy, algorithmName, buffer);
    dst3 = yflip(*dst3);

    if ((dst->getX0() != dst2->getX0()) || (dst->getX0() != dst3->getX0()) || (dst->getY0() != dst2->getY0()) || (dst->getY0() != dst3->getY0())) {
	cerr << s << "XY0 mismatch\n";
	return 1;
    }

    for (int i = 0; i < nx; i++) {
	for (int j = 0; j < ny; j++) {
	    // Note that offsetImage() is partly in single precision, so we test with 10^-6 threshold
	    if ((fabs((*dst)(i,j) - (*dst2)(i,j)) > 1.0e-6) || (fabs((*dst)(i,j) - (*dst3)(i,j)) > 1.0e-6)) {
		cerr << s << ": image mismatch\n";
#if 0
		cerr << "source image follows\n";
		show_image(src);
		cerr << "offset image follows\n";
		show_image(*dst);
		cerr << "xflip->offset->xflip image follows\n";
		show_image(*dst2);
		cerr << "yflip->offset->yflip image follows\n";
		show_image(*dst3);
#endif
		return 1;
	    }
	}
    }

    cerr << s << ": pass\n";
    return 0;
}


static int testFlipOffset()
{
    int n = 0;
    n += testOneFlipOffset(5, 5, 2, 2, 0.3, 0.2, "bilinear", 1);
    n += testOneFlipOffset(5, 5, 2, 2, -0.8, 0.3, "bilinear", 1);
    n += testOneFlipOffset(5, 5, 2, 2, -10.2, -8.2, "bilinear", 1);
    n += testOneFlipOffset(5, 5, 2, 2, 0.3, 0.2, "lanczos5", 5);
    n += testOneFlipOffset(5, 5, 2, 2, -0.8, 0.3, "lanczos5", 5);
    n += testOneFlipOffset(5, 5, 2, 2, -10.2, -8.2, "lanczos5", 5);
    return n;
}


// -------------------------------------------------------------------------------------------------
//
// a precise test of afw::math::offsetImage() with bilinear interpolation, by comparing with
// quick-and-dirty handcoded bilinear interpolation


// returns 0 on success, 1 on failure
static int testOneOffsetBilinear(int nx, int ny, int x0, int y0, double dx, double dy)
{
    // identifier string which will be displayed on failure
    ostringstream os;
    os << "testOneOffsetBilinear(" << nx << "," << ny << "," << x0 << "," << y0 << "," << dx << "," << dy << ")";
    string s = os.str();

    // initialize PSF to random image
    Image<double> src(nx, ny);
    src.setXY0(x0, y0);
    fill_random(src);

    Image<double>::Ptr dst = offsetImage(src, dx, dy, "bilinear", 1);

    // test correctness...
    // Note that offsetImage() is partly in single precision, so we test with 10^-6 threshold

    if ((dst->getWidth() != nx) || (dst->getHeight() != ny)) {
	cerr << s << ": mismatched dst dimensions\n";
	return 1;
    }

    int dst_x0 = dst->getX0();
    int dst_y0 = dst->getY0();

    int expected_dst_x0 = (fabs(dx) < 1.0) ? x0 : round(x0+dx);
    int expected_dst_y0 = (fabs(dy) < 1.0) ? y0 : round(y0+dy);

    if ((dst_x0 != expected_dst_x0) || (dst_y0 != expected_dst_y0)) {
	cerr << s << ": wrong image location\n";
	return 1;
    }

    for (int i = 0; i < nx; i++) {
	double si = i + dst_x0 - dx - x0;  // location in src image
	int si0 = floor(si);
	double sx = si-si0;

	for (int j = 0; j < ny; j++) {
	    double sj = j + dst_y0 - dy - y0;  // location in src image
	    int sj0 = floor(sj);
	    double sy = sj-sj0;
	    
	    // quick-and-dirty bilinear interpolation by hand
	    double t00 = (si0 >= 0 && sj0 >= 0) ? src(si0,sj0) : 0.0;
	    double t01 = (si0 >= 0 && sj0 < ny-1) ? src(si0,sj0+1) : 0.0;
	    double t10 = (si0 < nx-1 && sj0 >= 0) ? src(si0+1,sj0) : 0.0;
	    double t11 = (si0 < nx-1 && sj0 < ny-1) ? src(si0+1,sj0+1) : 0.0;
	    double t = t00*(1-sx)*(1-sy) + t01*(1-sx)*sy + t10*sx*(1-sy) + t11*sx*sy;
	    
	    if (fabs((*dst)(i,j)-t) > 1.0e-6) {
		cerr << s << ": incorrect output image at i=" << i << ", j=" << j << endl;
		cerr << "got " << (*dst)(i,j) << ", expected " << t << endl;
#if 0
		cerr << "t00=" << t00 << " t01=" << t01 << " t10=" << t10 << " t11=" << t11 << endl;
		cerr << "complete source image follows\n";
		show_image(src);
#endif
		return 1;
	    }
	}
    }

    cerr << s << ": pass\n";
    return 0;
}


static int testOffsetBilinear()
{
    int n = 0;
    n += testOneOffsetBilinear(5, 5, 2, 2, 10.0, 10.0);
    n += testOneOffsetBilinear(5, 5, 2, 2, 9.71, 9.65);
    n += testOneOffsetBilinear(5, 5, 2, 2, 9.9, 10.2);
    n += testOneOffsetBilinear(5, 5, 2, 2, 10.1, 9.8);
    n += testOneOffsetBilinear(5, 5, 2, 2, 10.1, 10.2);
    n += testOneOffsetBilinear(5, 5, 2, 2, 0.9, -0.8);
    return n;
}


// -------------------------------------------------------------------------------------------------


int main(int argc, char **argv)
{
    int n1 = testFlipOffset();
    if (n1 > 0)
	cerr << "testFlipOffset: " << n1 << " failures\n";

    int n2 = testOffsetBilinear();
    if (n2 > 0)
	cerr << "testOffsetBilinear: " << n2 << " failures\n";

    return (n1 > 0) || (n2 > 0);
}

/*
 * Local variables:
 *  c-basic-offset: 4
 * End:
 */
